autokeras/blocks/basic.py

```diff
@@ -260,6 +261,221 @@ def _get_padding(kernel_size, output_node):
+             inputs: Tensor of Shape [batch_size, seq_len, embedding_dim]
+
+        # Returns
+            Self-Attention outputs of shape `[batch_size, seq_len, embedding_dim]`.
+        """
+        inputs = nest.flatten(inputs)
+        utils.validate_num_inputs(inputs, 1)
+        input_node = inputs[0]
+        shape = input_node.shape.as_list()
+        if len(shape) != 3:
+            raise ValueError(
+                'Expect the input tensor to have '
+                '3 dimensions for multi-head self-attention, '
+                'but got {shape}'.format(shape=input_node.shape))
+        # input.shape = [batch_size, seq_len, embedding_dim]
+        embed_dim = self.embed_dim or hp.Choice(
+            'embed_dim',
+            [32, 64, 128, 256, 512],
+            default=128)
+        num_heads = self.num_heads
+        if num_heads is None:
+            num_heads = 8
+
+        if embed_dim % num_heads != 0:  # how to evaluate this condition
+            raise ValueError(
+                f"embedding dimension = {embed_dim} should be "
+                f"divisible by number of heads = {num_heads}"
+            )
+        projection_dim = embed_dim // num_heads
+        query_dense = layers.Dense(embed_dim)
+        key_dense = layers.Dense(embed_dim)
+        value_dense = layers.Dense(embed_dim)
+        combine_heads = layers.Dense(embed_dim)
+        batch_size = tf.shape(input_node)[0]
+        query = query_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        key = key_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        value = value_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        query, key, value = [self.separate_heads(
+            var, batch_size, num_heads,  projection_dim
+        ) for var in [query, key, value]]
+        attention, weights = self.attention(query, key, value)
+        attention = tf.transpose(
+            attention, perm=[0, 2, 1, 3]
+        )  # (batch_size, seq_len, num_heads, projection_dim)
+        concat_attention = tf.reshape(
+            attention, (batch_size, tf.shape(attention)[1], self.embed_dim)
+        )  # (batch_size, seq_len, embed_dim)
+        output = combine_heads(
+            concat_attention
+        )  # (batch_size, seq_len, embed_dim)
+        return output
+
+    @staticmethod
+    def attention(query, key, value):
+        score = tf.matmul(query, key, transpose_b=True)
+        dim_key = tf.cast(tf.shape(key)[-1], tf.float32)
+        scaled_score = score / tf.math.sqrt(dim_key)
+        weights = tf.nn.softmax(scaled_score, axis=-1)
+        output = tf.matmul(weights, value)
+        return output, weights
+
+    @staticmethod
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+class TransformerBlock(block_module.Block):
```

Comment:
Also please rename `TransformerBlock` to `Transformer`. We usually only add a "Block" as suffix whenever the name is confusing without it.